shotgun_api3/shotgun.py

```diff
@@ -2155,6 +2156,132 @@ def upload(self, entity_type, entity_id, path, field_name=None, display_name=Non
         is_thumbnail = (field_name in ["thumb_image", "filmstrip_thumb_image", "image",
                                        "filmstrip_image"])
 
+        # Thumbnails can only be uploaded to SG for now...
+        if self.server_info['s3_direct_uploads_enabled'] \
+                and entity_type == 'Version' and field_name == 'sg_uploaded_movie':
+            return self._upload_to_storage(entity_type, entity_id, path, field_name, display_name,
+                                           tag_list, is_thumbnail)
+        else:
+            return self._upload_to_sg(entity_type, entity_id, path, field_name, display_name,
+                                      tag_list, is_thumbnail)
+
+    def _upload_to_storage(self, entity_type, entity_id, path, field_name, display_name,
+                           tag_list, is_thumbnail):
+        """
+        Internal function to upload a file to S3 and link it to the specified entity.
```

Comment:
nitpick: given that we are being given an url from the server that we are uploading to, maybe we should just omit any mention of `S3`? AFAIK, as far as the shotgun API is concerned, it could be any sort of service and there is nothing S3 specific in the code right now other than comments and log messages.